File: core/src/main/java/gg/norisk/core/manager/InputbarPayloadManager.java
package gg.norisk.core.manager;

import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

import gg.norisk.core.payloads.AbstractInputbarPayload;
import gg.norisk.core.payloads.InputbarResponsePayload;
import gg.norisk.core.payloads.Payloads;

public class InputbarPayloadManager {
    private static final ConcurrentHashMap<UUID, InputSession> sessions = new ConcurrentHashMap<>();
    
    private static class InputSession {
        private final AbstractInputbarPayload payload;
        private final BiConsumer<UUID, byte[]> sendToClient;
        private final Consumer<String> handler;
        private final Runnable onCancel;
        
        public InputSession(AbstractInputbarPayload payload, BiConsumer<UUID, byte[]> sendToClient, 
                          Consumer<String> handler, Runnable onCancel) {
            this.payload = payload;
            this.sendToClient = sendToClient;
            this.handler = handler;
            this.onCancel = onCancel;
        }
        
        public AbstractInputbarPayload getPayload() {
            return payload;
        }
        
        public BiConsumer<UUID, byte[]> getSendToClient() {
            return sendToClient;
        }
        
        public Consumer<String> getHandler() {
            return handler;
        }
        
        public Runnable getOnCancel() {
            return onCancel;
        }
    }
    
    public static void registerInputSession(UUID playerUuid, AbstractInputbarPayload payload, 
                                          BiConsumer<UUID, byte[]> sendToClient, Consumer<String> handler, 
                                          Runnable onCancel) {
        sessions.put(playerUuid, new InputSession(payload, sendToClient, handler, onCancel));
        Payloads.send(playerUuid, payload, sendToClient);
    }
    
    public static void registerInputSession(UUID playerUuid, AbstractInputbarPayload payload, 
                                          BiConsumer<UUID, byte[]> sendToClient, Consumer<String> handler) {
        registerInputSession(playerUuid, payload, sendToClient, handler, null);
    }
    
    public static void unregisterSession(UUID playerUuid) {
        sessions.remove(playerUuid);
    }
    
    public static void handleInputbarResponse(UUID playerUuid, InputbarResponsePayload payload) {
        InputSession session = sessions.get(playerUuid);
        if (session == null) {
            return;
        }
        
        if (payload.isCanceled()) {
            if (session.getOnCancel() != null) {
                session.getOnCancel().run();
            }
        } else if (payload.getInput() != null) {
            session.getHandler().accept(payload.getInput());
            unregisterSession(playerUuid);
        }
    }
    
    public static void resendInputbar(UUID playerUuid) {
        InputSession session = sessions.get(playerUuid);
        if (session != null) {
            Payloads.send(playerUuid, session.getPayload(), session.getSendToClient());
        }
    }
} 

File: core/src/main/kotlin/gg/norisk/core/manager/InputbarPayloadManager.kt
package gg.norisk.core.manager

import gg.norisk.core.payloads.AbstractInputbarPayload
import gg.norisk.core.payloads.InputbarResponsePayload
import gg.norisk.core.payloads.Payloads
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

class InputbarPayloadManager {
    companion object {
        private data class InputSession(
            val payload: AbstractInputbarPayload,
            val sendToClient: (UUID, ByteArray) -> Unit,
            val handler: (String) -> Unit,
            val onCancel: (() -> Unit)? = null
        )

        private val sessions = ConcurrentHashMap<UUID, InputSession>()

        fun registerInputSession(
            playerUuid: UUID,
            payload: AbstractInputbarPayload,
            sendToClient: (UUID, ByteArray) -> Unit,
            handler: (String) -> Unit,
            onCancel: (() -> Unit)? = null
        ) {
            sessions[playerUuid] = InputSession(payload, sendToClient, handler, onCancel)
            Payloads.send(playerUuid, payload, sendToClient)
        }

        fun unregisterSession(playerUuid: UUID) {
            sessions.remove(playerUuid)
        }

        fun handleInputbarResponse(playerUuid: UUID, payload: InputbarResponsePayload) {
            val session = sessions[playerUuid] ?: run {
                return
            }

            if (payload.canceled) {
                session.onCancel?.invoke()
            } else if (payload.input != null) {
                session.handler(payload.input)
                unregisterSession(playerUuid)
            }
        }

        fun resendInputbar(playerUuid: UUID) {
            val session = sessions[playerUuid] ?: return
            Payloads.send(playerUuid, session.payload, session.sendToClient)
        }
    }
}